Python/Product/PythonTools/PythonTools/Intellisense/ProjectAnalyzer.cs

```diff
@@ -993,7 +1025,53 @@ SafePipeHandle stdInClientHandle
                 }
             ).ConfigureAwait(false);
 
-            if (response == null || response.fileId == -1) {
+            return FinishAnalyze(
+                response,
+                u => new AnalysisEntry(this, path, u, isTemporaryFile, suppressErrorList)
+            );
+        }
+
+        internal async Task<AnalysisEntry> AnalyzeFileAsync(
+            Uri uri,
+            bool isTemporaryFile = false,
+            bool suppressErrorList = false
+        ) {
+            if (_conn == null) {
+                // We aren't able to analyze code, so don't create an entry.
+                return null;
+            }
+
+            if (uri == null) {
+                throw new ArgumentNullException(nameof(uri));
+            }
+
+            AnalysisEntry entry;
+            if (_projectFilesByUri.TryGetValue(uri, out entry)) {
+                return entry;
+            }
+
+            string path = null;
+            if (uri.IsFile) {
+                path = uri.LocalPath;
+            }
+
+            Interlocked.Increment(ref _parsePending);
+            var response = await SendRequestAsync(
+                new AP.AddFileRequest {
+                    uri = uri,
+                    isTemporaryFile = isTemporaryFile,
+                    suppressErrorLists = suppressErrorList
+                }
+            ).ConfigureAwait(false);
+
+            return FinishAnalyze(
```

Comment:
Should `FinishAnalyze` be called in `finally` block?

Reply:
No. This is a straight refactor splitting one method into two.